src/include/miopen/tensor.hpp

```diff
@@ -173,6 +175,30 @@ struct TensorDescriptor : miopenTensorDescriptor
 
     std::string ToString() const;
 
+    template <class Vector, class Op>
+    inline std::vector<int64_t> sort_permutation(const Vector& data, Op op) const
+    {
+        std::vector<std::int64_t> result(data.size());
+        std::iota(result.begin(), result.end(), 0);
+        std::sort(
+            result.begin(), result.end(), [&](auto x, auto y) { return op(data[x], data[y]); });
+        return result;
+    }
+
+    std::string GetLayout(std::string labels) const
```

Comment:
Another consideration. When we support different layouts, layout comparisons (e.g. in `aSolver::IsApplicable()`) will happen quite often. Getting layouts is not very fast. Comparison of strings too. This could lead to some perf drop. Do you have a plan of optimization? Ideas?

Reply:
I see this in two ways:

- It does get more expensive, but in very limited manner. The 4-letter layout string is very easy to manipulate and adds only linear time complexity here. I don't think this is a big problem leaving it as-is.

- On the other hand, we can certainly cache it to a member variable (such that only the first time `GetLayout()` called that it is actually computed, otherwise it is just returning cached copies.). I can certainly implement this. Though I want to make sure you guys are okay with it in the sense that:
  1. It stores duplicated information against strides, as pointed out from original review comments.
  2. It is not very "functional", per say.

Reply:
> On the other hand, we can certainly cache it to a member variable (such that only the first time GetLayout() called that it is actually computed, otherwise it is just returning cached copies.). 

I dont think this is a good idea. `GetLayout`(or `ComputeLayout`) should be `const`. 

It would be better to store the layout string in the `ConvolutionContext` so that it only needs to be computed once.

Reply:
@pfultz2 Yah, think we mean the same, wasn't very clear about where to cache the layout string in my original comments. 

Addressed in latest commit.